### ISSUE TITLE
How to get TeX math font metrics?

### ISSUE THREAD

**Issue by F2X** [nodebox/opentype.js#107]
Hi, I would like to use [LatinModernMath-Regular](http://en.wikipedia.org/wiki/Computer_Modern#Latin_Modern) font to render math on a canvas.

I've used FontForge to inspect the *.otf file and I have a TeX entry with 3 sub-entries: TeX General, TeX Math Symbol and TeX Math Extension.
Is there a way to get these metrics using opentype.js?

**Comment by fdb** [comment]
Apparently these are [non-standard extensions used in FontForge](https://fontforge.github.io/non-standard.html). We don't support them. 

Since there aren't many TeX fonts, I suggest writing a plugin for FontForge that extracts the metrics data.

What do you want to use it for?

**Comment by F2X** [comment]
I would like to draw a math expression Glyph-by-glyph on a canvas.
I have a custom format representing a math expression, which is not MathML neither LaTeX, and I have to create a renderer for that. I'm currently in a research step/phase.

**Comment by fdb** [comment]
Unfortunately we don't have support for those tables, so you'll have to find another way. Python fontTools, which converts a TTF file to XML, might do the job as well, although I don't know if they have support for the TeX tables either.

**Comment by F2X** [comment]
Thanks, I will try it.

**Comment by moyogo** [comment]
fontTools can handle the MATH table. That table does the same thing as those non standard Fontforge TeX tables.

**Comment by F2X** [comment]
Yes, I saw that, it's probably more relevant to use these.
For now I'm using static values from FontForge for the font constants I need (e.g. fractionNumeratorShiftUp, Radicals values ...), but I'm always looking for something to do it dynamically, font independant

**Comment by Pomax** [comment]
a thought: table parsing plugins, where a plugin can say "I know how to parse a table with directionary tag `XYZ_` and can yield that table as object via parseTable and encode to binary using encodeTable" or something. That way OpenType.js itself doesn't need to get bogged down with nonstandard code, but we can offer an OpenType/addons, for instance, that comes with things like "MATH" parsing (provided someone is willing to write the parser object for it of course)

**Comment by davelab6** [comment]
:+1:

**Comment by fdb** [comment]
@Pomax I like this! However, it's not just about the tables. For example, sometimes there is a certain order in which we have to parse the table (e.g. `maxp` first to find out the number of glyphs). And once you have the data, how would it integrate with the higher-level `Font` and `Glyph` objects?

I think this is related to #67 and #72: providing a low-level table parsing / writing API, and a higher-level Glyph / Font API.

**Comment by Pomax** [comment]
hm, indicating the "required for a full parse" set should be fairly straight forward. A resolution callback should be relative easy: per loader indicate dependency tags in an array, and each time a table is loaded in, you splice that tag from the dependency list. Is the list empty? load this plugged-in table.

Also, there is a clear load order when it comes to non-standard tables: first the require tables (cmap, head, hhea, hmtx, maxp, name, OS/2, post), then the outline tables (at least, the table headers so that the pointers can be followed), then "anything else". So a dependency list would really only be necessary for non-required tables.

**Comment by fdb** [comment]
I agree. I was being a bit pedantic, thinking how a fully bootstrapped, everything-is-a-plugin approach could look. But that's perhaps taking it a bit too far...

**Comment by Pomax** [comment]
yeah keeping the true OT "approved" parsing the way it is makes a lot of sense (as non-negotiable, fixed set of tables). But if the additional table parsers turn out to be really to implement, thee's always the option to refactor =)
